src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy

```diff
@@ -0,0 +1,311 @@
+   * NamedDomainObjectContainer<PluginOptions>.
+   */
+  public void plugins(Closure configureClosure) {
+    ConfigureUtil.configure(configureClosure, plugins)
+  }
+
+  /**
+   * Add a directory to protoc's include path.
+   */
+  public void include(Object dir) {
+      if (dir instanceof File) {
+          includeDirs.add(dir)
+      } else {
+          includeDirs.add(project.file(dir))
+      }
+  }
+
+  /**
+   * The container of command-line options for a protoc plugin or a built-in output.
+   */
+  public static class PluginOptions implements Named {
+    private final ArrayList<String> options = new ArrayList<String>()
+    private final String name
+
+    public PluginOptions(String name) {
+      this.name = name
+    }
+
+    /**
+     * Adds a plugin option.
+     */
+    public PluginOptions option(String option) {
+      options.add(option)
+      return this
+    }
+
+    public List<String> getOptions() {
+      return options
+    }
+
+    /**
+     * Returns the name of the plugin or builtin.
+     */
+    @Override
+    public String getName() {
+      return name
+    }
+  }
+
+  //===========================================================================
+  //    protoc invocation logic
+  //===========================================================================
+
+  // protoc allows you to prefix comma-delimited options to the path in
+  // the --*_out flags, e.g.,
+  // - Without options: --java_out=/path/to/output
+  // - With options: --java_out=option1,option2:/path/to/output
+  // This method generates the prefix out of the given options.
+  def String makeOptionsPrefix(List<String> options) {
+    StringBuilder prefix = new StringBuilder()
+    if (!options.isEmpty()) {
+      options.each { option ->
+        if (prefix.length() > 0) {
+          prefix.append(',')
+        }
+        prefix.append(option)
+      }
+      prefix.append(':')
+    }
+    return prefix.toString()
+  }
+
+  String getOutputDir(PluginOptions plugin) {
+    return "${outputBaseDir}/${plugin.name}"
+  }
+
+  Collection<File> getAllOutputDirs() {
+    ImmutableList.Builder<File> dirs = ImmutableList.builder()
+    builtins.each { builtin ->
+      dirs.add(new File(getOutputDir(builtin)))
+    }
+    plugins.each { plugin ->
+      dirs.add(new File(getOutputDir(plugin)))
+    }
+    return dirs.build()
+  }
+
+  @TaskAction
+  def compile() {
+    Preconditions.checkState(!initializing, 'doneInitializing() has not been called')
+
+    // Only at this point all outputs are finalized, then we can register the
+    // generated source dirs to java compilation.
+
+    // NOTE(zhangkun83): does this mean we can actually allow per-plugin output
+    // dir reconfiguraiton? Not actually -- the basedir of all outputs must be
+    // registered to task.outputs before Gradle starts to execute tasks, so
+    // that the task can be correctly skipped.  We may only allow plugins to
```

Comment:
I thought task dependency graph was created during evaluation, but [it seems](https://docs.gradle.org/current/userguide/build_lifecycle.html) dependency is calculated after evaluation, then this should be fine. I will update the comment to include a TODO instead.